### PR TITLE
Some additions desired by RP-0

### PR DESCRIPTION
**Pull request by NathanKell**
1. Support external checking of whether a crew member can be placed in a part (not tied to RP-0 per se).
2. Fake-unlock all R&D nodes in the techlist when entering the R&D screen. Nodes currently being researched will have orange icons but still green backgrounds.
3. Add option where a pad can't have something rolled out to it if it's currently being reconditioned.
4. Show rollout cost next to the rollout button on the KCT spacecenter GUI VAB tab.
5. Support global variables for BP cost: total craft BP is multiplied by that var if >= 1 part has the module named.
6. Supports a limit to max rush build clicks (note: not settable in the GUI).
7. Tracks vessel number of stages, number of staged parts, and cost of staged parts. Adds new variables to that effect.

### CODE REVIEW

**Comment by magico13** [comment]
Are you ready for these to be merged in? I know you were still adding things for a little while. I'm checking them over now but doubt there will be anything I won't be fine with.

File: Kerbal_Construction_Time/KCT_GUI_Presets.cs [hunk 1]
```diff
@@ -113,6 +113,7 @@ public static void DrawPresetWindow(int windowID)
             WorkingPreset.generalSettings.Enabled= GUILayout.Toggle(WorkingPreset.generalSettings.Enabled, "Mod Enabled", HighLogic.Skin.button);
             WorkingPreset.generalSettings.BuildTimes = GUILayout.Toggle(WorkingPreset.generalSettings.BuildTimes, "Build Times", HighLogic.Skin.button);
             WorkingPreset.generalSettings.ReconditioningTimes = GUILayout.Toggle(WorkingPreset.generalSettings.ReconditioningTimes, "Launchpad Reconditioning", HighLogic.Skin.button);
+            WorkingPreset.generalSettings.ReconditioningTimes = GUILayout.Toggle(WorkingPreset.generalSettings.ReconditioningBlocksPad, "Reconditioning Blocks Pad", HighLogic.Skin.button);
```
Comment (magico13) [inline Kerbal_Construction_Time/KCT_GUI_Presets.cs]: Not sure if you caught this later, but that's probably not right.

**Comment by NathanKell** [comment]
Haha whoops! Thanks! :)

Not quite ready--there's an issue where it's complaining about how I added the new variables, that they aren't doubles. Do I need to cast or something? Also we're not sure how else to handle trying to detect vessel complexity, because 4 ullage motors won't add the same complexity as 4 F-1 engines, let alone having RCS activate due to a stage rather than hitting the R key really should not be increasing perceived complexity.

I also didn't make a UI element for setting the number of rushbuild clicks. Not sure if you care about that. :]

Other than that it's all been working fine in testing so far AFAIK so I'm fairly confident of the PR. And I'm not sure what other small stuff there is to add.

**Comment by magico13** [comment]
Sorry, meant to reply to this sooner. I'll pull it in as soon as you give the all clear.

> there's an issue where it's complaining about how I added the new variables, that they aren't doubles. Do I need to cast or something?

They get passed in as strings, but they have to be parseable to doubles. MagiCore should print out the current state of the formula so you should be able to tell pretty quickly if something isn't going to parse correctly. Feel free to send the log my way if you think it should be working but isn't.

**Comment by NathanKell** [comment]
Hah, no worries, turns out I *also* forgot to reply here and said I resolved the problem! :D (The issue was I was using those variables in another formula but had forgotten to add the code so they were in the dict for that formula).

I do however have a couple remaining questions that you might have insight on:

1. The rollout cost shown in the editor's KCT window doesn't seem to be related to the rollout cost shown/paid in spacecenter view when actually rolling out.
2. When starting a new game, in the select preset/do upgrades bit, I select the RP-0 preset and then go to upgrades, but the build rates are wrong (VAB1 shows 0.1BP/s, the button shows +0.05, and VAB2 is visible) but once you add a point to VAB1 it clears up (we changed so it should start at 0.05 and 1pt raises to 0.0525, and VAB2 is locked until VAB upgrade). Any ideas?

**Comment by magico13** [comment]
I know the latter is already an issue (#119). I'm pretty sure it's just a caching issue and the cached values need to be reset when saving. I haven't really looked into it yet but I probably can tonight.

The rollout cost is split up into tenths along the rollout process. The whole thing should add up to the total cost, but might not if it advances quickly. Please feel free to change that if you'd like since I know of nobody else with plans to use that. If you want to make it just an upfront cost, or a cost at the end of rollout, then that's 100% cool with me. In the development branch that code is around here: https://github.com/magico13/KCT/blob/development/Kerbal_Construction_Time/KCT_Utilities.cs#L493

**Comment by NathanKell** [comment]
Ah, I wasn't clear about costs then. It's not that. It's that, I think, the editor GUI is using a different pad level than what KCT currently has as the active pad. For example, the editor KCT GUI says something like 7000 funds to roll out, but in practice it's 3000.

(Recall I added a display of rollout cost to that pad, in the VAB tab of the KCT build list in spacecenter, so I know what rollout cost I'll be paying for that pad level.)

**Comment by magico13** [comment]
Hmm, that'd be pretty funky if it was giving a different level, since the level is still gotten the same way... I believe I fixed something with getting the level last night, but it shouldn't affect launchpads since those are pulled straight from KCT's info and not from KSP. You could put a breakpoint on the KCT_MathParsing.ParseRolloutCost method (or just write to the log) to see what the variables are and check that they are indeed different launchpad levels being passed.

You might just replace KCT_GameStates.EditorRolloutCosts in [KCT_GUI](https://github.com/KSP-RO/KCT/blob/078da77027360f15b1d3fcb4c1b82d6372ded83c/Kerbal_Construction_Time/KCT_GUI.cs#L420) with a call to KCT_MathParsing.ParseRolloutCost, but it should already be updated automatically whenever the vessel is altered in any way during editing, so it likely won't make a difference.

Are the Editor UI and the Build List really giving different results? I'd just check the variables to see what's different, since they should be the same for the same vessel.

**Comment by NathanKell** [comment]
Am back. Am retargeting for 1.3. Would you like to make a PR to main, or do you want to merge this and then merge up?

**Comment by NathanKell** [comment]
Never mind. I can't read the next line up. Oy. (And thanks @aw1621107 ! )

**Pull request closed without merging**